### PR TITLE
`Quiz exercises`: Allow file editing when importing from overview

### PR DESCRIPTION
**Pull request by EneaGore**
#### General

- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
~- [ ] I added multiple integration tests (Spring) related to the features (with a high test coverage).~
~- [ ] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).~
- [x] I documented the Java code using JavaDoc style.

#### Client
No changes

### Motivation and Context

When making changes to a quiz that the user imports through "Import Quiz" from the exercise overview, adding new files throws an internal server error. This is because the new files are sent to the server, but nothing was done with them except to be  validated and throw an error ...

### Description

When the exercises were imported, the system checked if the file paths for the background and drag items were correct upload Paths. Since new images obviously do not have a correct path when they are new, I added a check for this. Then, I created a method that saves these images and links the path to the background/drag item path of the exercises.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Quiz

1. Navigate to Course Administration -> Exercises
2. From the overview press on "Import Quiz"
3. Make changes, specifically add/remove files
4. Save
5. No server error should be shown
6. Check that new quiz should have all changes saved
7. Check that quiz you imported from has not been changed

#### Exam Mode Testing

Nothing should have changed from the exam exercise group import.

[](https://artemis-test1.artemis.cit.tum.de)
[](https://artemis-test2.artemis.cit.tum.de)
[](https://artemis-test3.artemis.cit.tum.de)
[](https://artemis-test4.artemis.cit.tum.de)
[](https://artemis-test5.artemis.cit.tum.de)
[](https://artemis-test6.artemis.cit.tum.de)
[](https://artemis-test9.artemis.cit.tum.de)

#### Performance Review
- [x] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

## Summary by CodeRabbit

- **New Features**
	- Enhanced the import process for drag-and-drop questions to include checks for valid file paths, improving reliability.
	- Introduced a new method to facilitate the upload of new files associated with quiz exercises, specifically for drag-and-drop questions.

- **Bug Fixes**
	- Improved error handling by adding support for IOException in the import process, ensuring robustness during file operations.

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/service/quiz/QuizExerciseService.java [hunk 1]
```diff
@@ -458,4 +458,38 @@ protected QuizExercise saveAndFlush(QuizExercise quizExercise) {
...
+
+    /**
+     *
+     * @param newQuizExercise the newly created quiz exercise, after importing basis of imported exercise
+     * @param files           the new files to be added to the newQuizExercise which do not have a previous path and need to be saved in the server
+     * @return the new exercise with the updated file paths which have been created and saved
+     * @throws IOException
+     */
+    public QuizExercise uploadNewFilesToNewImportedQuiz(QuizExercise newQuizExercise, List<MultipartFile> files) throws IOException {
+        Map<String, MultipartFile> fileMap = files.stream().collect(Collectors.toMap(MultipartFile::getOriginalFilename, file -> file));
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/service/quiz/QuizExerciseService.java:470]:
```suggestion
        Map<String, MultipartFile> fileMap = files.stream().collect(Collectors.toMap(MultipartFile::getOriginalFilename, Function.identity()));
```
Makes it a bit clearer that the file stays unchanged

File: src/main/java/de/tum/in/www1/artemis/service/quiz/QuizExerciseService.java [hunk 2]
```diff
@@ -458,4 +458,38 @@ protected QuizExercise saveAndFlush(QuizExercise quizExercise) {
...
+                    if (dragItem.getPictureFilePath() != null) {
+                        String filePathDrag = dragItem.getPictureFilePath();
+                        MultipartFile dragFile = fileMap.get(filePathDrag);
+                        URI publicPathUriDrag = URI.create(filePathDrag);
+                        if (FilePathService.actualPathForPublicPath(publicPathUriDrag) == null) {
+                            String newFilePath = saveDragAndDropImage(FilePathService.getDragItemFilePath(), dragFile, dragItem.getId()).toString();
+                            dragItem.setPictureFilePath(newFilePath);
+                        }
+                    }
+                }
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/service/quiz/QuizExerciseService.java:490]: Can you move this into another method, so we do not end up with 5 or 6 indentions?

File: src/main/java/de/tum/in/www1/artemis/web/rest/QuizExerciseResource.java [hunk 3]
```diff
@@ -792,8 +792,10 @@ public ResponseEntity<QuizExercise> importExercise(@PathVariable long sourceExer
         quizExerciseService.validateQuizExerciseFiles(importedExercise, nullsafeFiles, false);
 
         final var originalQuizExercise = quizExerciseRepository.findByIdElseThrow(sourceExerciseId);
-        final var newQuizExercise = quizExerciseImportService.importQuizExercise(originalQuizExercise, importedExercise);
-
+        QuizExercise newQuizExercise = quizExerciseImportService.importQuizExercise(originalQuizExercise, importedExercise);
+        if (files != null) {
+            newQuizExercise = quizExerciseService.save(quizExerciseService.uploadNewFilesToNewImportedQuiz(newQuizExercise, files));
+        }
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/web/rest/QuizExerciseResource.java:798]: Can you move the file import into the service method as well?

**Review by JohannesStoehr — CHANGES_REQUESTED** [review]
Code looks good overall, left a few smaller comments:

File: src/main/java/de/tum/in/www1/artemis/service/quiz/QuizExerciseService.java (hunk 2 above)
Comment (EneaGore) [reply src/main/java/de/tum/in/www1/artemis/service/quiz/QuizExerciseService.java]: merged 2 if statement in one condition, slightly modified existing methods from this class and reused them.

**Review by N0W0RK — APPROVED** [review]
Code

**Review by JohannesStoehr — APPROVED** [review]
Code

**Review by BBesrour — APPROVED** [review]
Tested on TS5, works as expected

File: src/main/java/de/tum/in/www1/artemis/service/quiz/QuizExerciseImportService.java [hunk 4]
```diff
@@ -71,10 +75,11 @@ public QuizExerciseImportService(QuizExerciseService quizExerciseService, FileSe
      *
      * @param templateExercise The template exercise which should get imported
      * @param importedExercise The new exercise already containing values which should not get copied, i.e. overwritten
+     * @param files            The potential files to be added. Null if no change to files during import. ExamImportService sends null by default
      * @return The newly created exercise
      */
     @NotNull
-    public QuizExercise importQuizExercise(final QuizExercise templateExercise, QuizExercise importedExercise) {
+    public QuizExercise importQuizExercise(final QuizExercise templateExercise, QuizExercise importedExercise, @Nullable List<MultipartFile> files) throws IOException {
```
Comment (ole-ve) [inline src/main/java/de/tum/in/www1/artemis/service/quiz/QuizExerciseImportService.java:82]: Is this just affecting new files, i.e. existing files won't be changed?

If yes, I would probably overload the method as setting it to null essentially means "no additional effect".

**Review by laadvo — APPROVED** [review]
tested on ts4, quiz importing is working as expected

File: src/main/java/de/tum/in/www1/artemis/service/quiz/QuizExerciseImportService.java (hunk 4 above)
Comment (EneaGore) [reply src/main/java/de/tum/in/www1/artemis/service/quiz/QuizExerciseImportService.java:82]: I get what you mean, but this method is only used twice and I dont think there is any clear benefit in this instance to overload the method.

**Pull request merged**